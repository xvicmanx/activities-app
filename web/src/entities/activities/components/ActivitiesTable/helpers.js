// @flow

import moment from 'moment';

export const dateValueTableValueResolver = (item: Object): any => moment(item.date).format('DD-MM-YYYY HH:mm');

