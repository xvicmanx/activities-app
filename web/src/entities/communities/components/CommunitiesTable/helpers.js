// @flow

import React from 'react';
import { Link } from 'react-router-dom';

export const nameTableValueResolver = (item: Object): React$Element<any> => (
  <Link to={`/communities/${item.id}`}>{item.name}</Link>
);
