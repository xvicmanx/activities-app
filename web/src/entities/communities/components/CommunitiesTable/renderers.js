// @flow

import React from 'react';

type RendererProps = {
  field: Object,
};

export const DescriptionRenderer = ({
  field,
}: RendererProps): React$Element<any> => <textarea {...field} />;
