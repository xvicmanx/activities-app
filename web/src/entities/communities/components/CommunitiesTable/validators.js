export const createFormValidator = (values) => {
  const errors = {};

  if (!values.name) {
    errors.name = 'Por favor, provea un nombre';
  }

  if (!values.slogan) {
    errors.slogan = 'Por favor, provea un eslogan';
  }

  return errors;
};

export const updateFormValidator = (values) => {
  const errors = {};

  if (!values.id) {
    errors.id = 'Por favor, provea el id';
  }

  if (!values.name) {
    errors.name = 'Por favor, provea un nombre';
  }

  if (!values.slogan) {
    errors.slogan = 'Por favor, provea un eslogan';
  }

  return errors;
};

export const deleteFormValidator = (values) => {
  const errors = {};

  if (!values.id) {
    errors.id = 'Por favor, provea el id';
  }

  return errors;
};
