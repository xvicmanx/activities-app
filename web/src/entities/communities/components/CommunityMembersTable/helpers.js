// @flow

export const coordinatesTableValueResolver = (item: Object): string =>
  item.coordinates ? 'Si' : 'No';
