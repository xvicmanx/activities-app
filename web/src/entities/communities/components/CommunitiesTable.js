// @flow

import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import CRUDTable, {
  CreateForm,
  DeleteForm,
  Fields,
  Field,
  Pagination,
  UpdateForm,
} from 'react-crud-table';

import { readTokenFromCookie } from '../../users/redux/UsersActions';
import CommunitiesService from '../services/CommunitiesService';
import { DEFAULT_OPTIONS, DEFAULT_RESULT, encode } from '../../../core/helpers';
import type { Options } from '../../../core/helpers';

type RendererProps = {
  field: Object,
};

const DescriptionRenderer = ({ field }: RendererProps) => (
  <textarea {...field} />
);


const service = {
  fetchItems: async (options: Options) => {
    const response = await CommunitiesService.fetchCommunities(
      readTokenFromCookie(),
      encode(options),
    );
    return {
      items: response.communities,
      total: response.total,
    };
  },
  create: (community) =>
    CommunitiesService.createCommunity(community, readTokenFromCookie()),
  update: (community) =>
    CommunitiesService.updateCommunity(community, readTokenFromCookie()),
  delete: (community) =>
    CommunitiesService.deleteCommunity(community.id, readTokenFromCookie()),
};

const CommunitiesTable = (): React$Element<any> => {
  const [options, setOptions] = useState(DEFAULT_OPTIONS);
  const [result, setResult] = useState(DEFAULT_RESULT);

  useEffect(() => {
    const fetchItems = async () => {
      const res = await service.fetchItems(options);
      setResult(res);
    };
    fetchItems();
  }, [options]);

  return (
    <CRUDTable
      caption="Comunidades"
      items={result.items}
      actionsLabel="Acciones"
      onChange={(data) => {
        setOptions({
          sort: data.sort,
          queryRules: data.queryRules,
          activePage: data.pagination.activePage,
          itemsPerPage: data.pagination.itemsPerPage,
        });
      }}
    >
      <Fields>
        <Field name="id" label="Id" hideInCreateForm readOnly />
        <Field
          name="name"
          label="Nombre"
          placeholder="Nombre"
          tableValueResolver={(item) => (
            <Link to={`/communities/${item.id}`}>{item.name}</Link>
          )}
        />
        <Field name="slogan" label="Eslogan" render={DescriptionRenderer} />
      </Fields>
      <CreateForm
        title="Crear comunidad"
        message="Crear una nueva comunidad"
        trigger="Crear comunidad"
        onSubmit={(task) => service.create(task)}
        submitText="Crear"
        validate={(values) => {
          const errors = {};
  
          if (!values.name) {
            errors.name = 'Por favor, provea un nombre';
          }
  
          if (!values.slogan) {
            errors.slogan = 'Por favor, provea un eslogan';
          }
  
          return errors;
        }}
      />
      <UpdateForm
        title="Actualizar comunidad"
        message="Actualizar comunidad"
        trigger="Actualizar"
        onSubmit={service.update}
        submitText="Actualizar"
        validate={(values) => {
          const errors = {};
  
          if (!values.id) {
            errors.id = 'Por favor, provea el id';
          }
  
          if (!values.name) {
            errors.name = 'Por favor, provea un nombre';
          }
  
          if (!values.slogan) {
            errors.slogan = 'Por favor, provea un eslogan';
          }
  
          return errors;
        }}
      />
      <DeleteForm
        title="Eliminar comunidad"
        message="Esta seguro que quiere eliminar la comunidad?"
        trigger="Eliminar"
        onSubmit={service.delete}
        submitText="Eliminar"
        validate={(values) => {
          const errors = {};
  
          if (!values.id) {
            errors.id = 'Por favor, provea el id';
          }
  
          return errors;
        }}
      />
      <Pagination
        defaultActivePage={1}
        itemsPerPage={options.itemsPerPage}
        totalOfItems={result.total}
      />
    </CRUDTable>
  );
}


export default CommunitiesTable;
