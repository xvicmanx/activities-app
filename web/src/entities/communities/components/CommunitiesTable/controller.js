// @flow

import { encode } from '../../../../core/helpers';
import type { Options } from '../../../../core/helpers';
import { readToken } from '../../../users/redux/UsersActions';
import CommunitiesService from '../../services/CommunitiesService';

const Controller = {
  fetchItems: async (
    options: Options
  ): Promise<{ items: any, total: any, ... }> => {
    const response = await CommunitiesService.fetchCommunities(
      readToken(),
      encode(options)
    );
    return {
      items: response.communities,
      total: response.total,
    };
  },
  create: async (community: Object): Promise<any> => {
    try {
      const result = await CommunitiesService.createCommunity(
        community,
        readToken()
      );
      return result;
    } catch (error) {
      return {
        success: false,
        message: error.message,
      };
    }
  },
  update: async (community: Object): Promise<any> => {
    try {
      const result = await CommunitiesService.updateCommunity(
        community,
        readToken()
      );
      return result;
    } catch (error) {
      return {
        success: false,
        message: error.message,
      };
    }
  },
  delete: async (community: Object): Promise<any> => {
    try {
      const result = await CommunitiesService.deleteCommunity(
        community.id,
        readToken()
      );
      return result;
    } catch (error) {
      return {
        success: false,
        message: error.message,
      };
    }
  },
};

export default Controller;
