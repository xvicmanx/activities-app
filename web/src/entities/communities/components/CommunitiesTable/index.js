// @flow

import React, { useState, useEffect } from 'react';
import CRUDTable, {
  CreateForm,
  DeleteForm,
  Fields,
  Field,
  Pagination,
  UpdateForm,
} from 'react-crud-table';

import {
  DEFAULT_OPTIONS,
  DEFAULT_RESULT,
  throwErrorWhenNotSuccess,
} from '../../../../core/helpers';
import {
  createFormValidator,
  deleteFormValidator,
  updateFormValidator,
} from './validators';
import { DescriptionRenderer } from './renderers';
import Controller from './controller';
import { nameTableValueResolver } from './helpers';

const CommunitiesTable = (): React$Element<any> => {
  const [options, setOptions] = useState(DEFAULT_OPTIONS);
  const [result, setResult] = useState(DEFAULT_RESULT);

  useEffect(() => {
    const fetchItems = async () => {
      const res = await Controller.fetchItems(options);
      setResult(res);
    };
    fetchItems();
  }, [options]);

  return (
    <CRUDTable
      caption="Comunidades"
      items={result.items}
      actionsLabel="Acciones"
      showQueryBuilder
      onChange={(data) => {
        setOptions({
          sort: data.sort,
          queryRules: data.queryRules,
          activePage: data.pagination.activePage,
          itemsPerPage: data.pagination.itemsPerPage,
        });
      }}
    >
      <Fields>
        <Field name="id" label="Id" hideInCreateForm readOnly />
        <Field
          name="name"
          label="Nombre"
          placeholder="Nombre"
          tableValueResolver={nameTableValueResolver}
        />
        <Field name="slogan" label="Eslogan" render={DescriptionRenderer} />
      </Fields>
      <CreateForm
        title="Crear comunidad"
        message="Crear una nueva comunidad"
        trigger="Crear comunidad"
        onSubmit={throwErrorWhenNotSuccess(Controller.create)}
        submitText="Crear"
        validate={createFormValidator}
      />
      <UpdateForm
        title="Actualizar comunidad"
        message="Actualizar comunidad"
        trigger="Actualizar"
        onSubmit={throwErrorWhenNotSuccess(Controller.update)}
        submitText="Actualizar"
        validate={updateFormValidator}
      />
      <DeleteForm
        title="Eliminar comunidad"
        message="Esta seguro que quiere eliminar la comunidad?"
        trigger="Eliminar"
        onSubmit={throwErrorWhenNotSuccess(Controller.delete)}
        submitText="Eliminar"
        validate={deleteFormValidator}
      />
      <Pagination
        defaultActivePage={1}
        itemsPerPage={options.itemsPerPage}
        totalOfItems={result.total}
      />
    </CRUDTable>
  );
};

export default CommunitiesTable;
