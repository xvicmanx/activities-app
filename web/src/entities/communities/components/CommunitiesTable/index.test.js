import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { BrowserRouter as Router } from 'react-router-dom';
import { act } from 'react-dom/test-utils';

import { DEFAULT_OPTIONS } from '../../../../core/helpers';
import CommunitiesTable from './';
import Controller from './controller';

jest.mock('../../../../core/useTableItemsFetch');

import useTableItemsFetch from '../../../../core/useTableItemsFetch';

Enzyme.configure({ adapter: new Adapter() });

describe('CommunitiesTable', () => {
  const communities = [
    {
      id: 1,
      name: 'Test Community 1',
    },
    {
      id: 2,
      name: 'Test Community 2',
    },
  ];
  let itemsResolver;

  beforeEach(() => {
    useTableItemsFetch.mockClear()
      .mockImplementation(() => ({
        total: communities.length,
        items: communities,
      }));
  });

  it('renders properly', async () => {
    const result = Enzyme.mount(<Router><CommunitiesTable /></Router>);

    expect(useTableItemsFetch).toHaveBeenCalledTimes(1);
    expect(useTableItemsFetch).toHaveBeenCalledWith(
      Controller.fetchItems,
      DEFAULT_OPTIONS,
    );
    expect(result.find('CRUDTable').props()).toMatchSnapshot();
  });

  it('updates the options on change', async () => {
    const result = Enzyme.mount(<Router><CommunitiesTable /></Router>);
    const data = {
      sort: {
        field: 'name',
        direction: 'descending', 
      },
      queryRules: [],
      pagination: {
        activePage: 2,
        itemsPerPage: 15,
      },
    };

    result.find('CRUDTable').prop('onChange')(data);
    result.update();
    expect(result.find('CRUDTable').props()).toMatchSnapshot();
  });
});
