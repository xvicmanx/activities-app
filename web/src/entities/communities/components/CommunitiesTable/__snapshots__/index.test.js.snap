// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommunitiesTable renders properly 1`] = `
Object {
  "actionsLabel": "Acciones",
  "caption": "Comunidades",
  "children": Array [
    <CRUDTable_Fields>
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={true}
        hideInUpdateForm={false}
        label="Id"
        name="id"
        queryable={true}
        readOnly={true}
        sortable={true}
        tableValueResolver={null}
        type="text"
      />
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={false}
        hideInUpdateForm={false}
        label="Nombre"
        name="name"
        placeholder="Nombre"
        queryable={true}
        sortable={true}
        tableValueResolver={[Function]}
        type="text"
      />
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={false}
        hideInUpdateForm={false}
        label="Eslogan"
        name="slogan"
        queryable={true}
        render={[Function]}
        sortable={true}
        tableValueResolver={null}
        type="text"
      />
    </CRUDTable_Fields>,
    <CRUDTable_CreateForm
      message="Crear una nueva comunidad"
      onSubmit={[Function]}
      submitText="Crear"
      title="Crear comunidad"
      trigger="Crear comunidad"
      validate={[Function]}
    />,
    <CRUDTable_UpdateForm
      message="Actualizar comunidad"
      onSubmit={[Function]}
      submitText="Actualizar"
      title="Actualizar comunidad"
      trigger="Actualizar"
      validate={[Function]}
    />,
    <CRUDTable_DeleteForm
      message="Esta seguro que quiere eliminar la comunidad?"
      onSubmit={[Function]}
      submitText="Eliminar"
      title="Eliminar comunidad"
      trigger="Eliminar"
      validate={[Function]}
    />,
    <CRUDTable_Pagination
      defaultActivePage={1}
      itemsPerPage={20}
      totalOfItems={2}
    />,
  ],
  "fetchItems": null,
  "items": Array [
    Object {
      "id": 1,
      "name": "Test Community 1",
    },
    Object {
      "id": 2,
      "name": "Test Community 2",
    },
  ],
  "onChange": [Function],
  "showQueryBuilder": true,
}
`;

exports[`CommunitiesTable updates the options on change 1`] = `
Object {
  "actionsLabel": "Acciones",
  "caption": "Comunidades",
  "children": Array [
    <CRUDTable_Fields>
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={true}
        hideInUpdateForm={false}
        label="Id"
        name="id"
        queryable={true}
        readOnly={true}
        sortable={true}
        tableValueResolver={null}
        type="text"
      />
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={false}
        hideInUpdateForm={false}
        label="Nombre"
        name="name"
        placeholder="Nombre"
        queryable={true}
        sortable={true}
        tableValueResolver={[Function]}
        type="text"
      />
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={false}
        hideInUpdateForm={false}
        label="Eslogan"
        name="slogan"
        queryable={true}
        render={[Function]}
        sortable={true}
        tableValueResolver={null}
        type="text"
      />
    </CRUDTable_Fields>,
    <CRUDTable_CreateForm
      message="Crear una nueva comunidad"
      onSubmit={[Function]}
      submitText="Crear"
      title="Crear comunidad"
      trigger="Crear comunidad"
      validate={[Function]}
    />,
    <CRUDTable_UpdateForm
      message="Actualizar comunidad"
      onSubmit={[Function]}
      submitText="Actualizar"
      title="Actualizar comunidad"
      trigger="Actualizar"
      validate={[Function]}
    />,
    <CRUDTable_DeleteForm
      message="Esta seguro que quiere eliminar la comunidad?"
      onSubmit={[Function]}
      submitText="Eliminar"
      title="Eliminar comunidad"
      trigger="Eliminar"
      validate={[Function]}
    />,
    <CRUDTable_Pagination
      defaultActivePage={1}
      itemsPerPage={15}
      totalOfItems={2}
    />,
  ],
  "fetchItems": null,
  "items": Array [
    Object {
      "id": 1,
      "name": "Test Community 1",
    },
    Object {
      "id": 2,
      "name": "Test Community 2",
    },
  ],
  "onChange": [Function],
  "showQueryBuilder": true,
}
`;
